brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+Most language-specific implementations will present SOMADataFrame in more convenient forms, such as Python `pandas.DataFrame`, R `data.frame` or other environment-specific data structures. Specific implementations of the SOMA API may extend this type with convertors to/from the underlying data structure, and enforce indexing conventions.
+
+> ⚠️ **Issue** - this spec requires that index column is a `string` type. Most underlying "engines" will support more flexibility. Do we want to enable that flexibility, or stick with requiring string for simplicity/commonality?
+
+## SOMAMatrix
+
+SOMAMatrix is a 2-D multi-column table of monomorphic arrays, ie, a "dataframe" with two index dimensions. SOMAMatrix is an extension of SOMADataFrame, with two indexing dimensions. The SOMAMatrix has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining the number of columns, and their respective names and types
+- the _primary_ index column name
+- the _secondary_ index column name
+
+A SOMAMatrix must contain (minimally) two columns which are the primary and secondary index columns. The index columns _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+> ⚠️ **Issue** - SOMAMatrix and SOMADataFrame differ only in the number of dimensions (2 vs 1 respectively). Should they be unified into a single type object/API of more general nature? Clarity vs flexibility?
+
+## SOMA
+
+The SOMA type is a specialized SOMACollection, representing an annotated 2-D matrix of observations across a set of variables. The SOMA composes SOMACollection, SOMADataFrame and SOMAMatrix types using a set of predefined fields (keys & values), to represent the annotated matrix. Each predefined field in the SOMA has a well-defined type, dimensionality and indexing convention. Other user defined data may be added to a SOMA, as a SOMA is a specialization of a SOMACollection.
+
+Implementations _should_ enforce the naming and indexing constraints on these pre-defined fields. Pre-defined fields are distinguished from other user-defined collection elements, where no schema or indexing semantics are presumed or enforced.
+
+The predefined fields within SOMA share common indices and `shape`:
+
+- obs_id - the observation, a `string` type
+- var_id - the variable, a `string` type
+
+SOMA pre-defined fields are principally differentiated by:
+
+- dimensionality, eg, 1-D (SOMADataFrame) or 2-D (SOMAMatrix)
+- index column names, `obs_id` or `var_id` or both
```

Comment:
if we're being specific about spelling `obs_id` and `var_id` then we may as well add `obs_id_i`/`obs_od_j` and `var_id_i`/`var_id_j` too

Reply:
this is specified in the constraints section.